src/main/surfrdMod.F90

```diff
@@ -882,6 +887,115 @@ subroutine surfrd_veg_dgvm(begg, endg)
+    call ncd_inqdid(ncid,'nmaxhillcol',dimid,readvar) 
+    if (.not. readvar) then
+       call endrun( msg=' ERROR: nmaxhillcol not on surface data file'//errMsg(sourcefile, __LINE__))
+    else
+       call ncd_inqdlen(ncid,dimid,nh)
+       max_columns_hillslope = nh
+    endif
+    ! actual number of columns per landunit
+    allocate(arrayl(begg:endg))
+    call ncd_io(ncid=ncid, varname='nhillcolumns', flag='read', data=arrayl, &
+         dim1name=grlnd, readvar=readvar)
+    if (.not. readvar) then
+       call endrun( msg=' ERROR: nhillcolumns not on surface data file'//errMsg(sourcefile, __LINE__))
+    else
+       ncolumns_hillslope(begg:endg) = arrayl(begg:endg)
+    endif
+    deallocate(arrayl)
+
+    ! pft_from_file and pft_lowland_upland assume that 1 pft
+    ! will exist on each hillslope column.  In prepration, set one
+    ! pft weight to 100 and the rest to 0.  The vegetation type
+    ! (patch%itype) will be reassigned when initHillslope is called later.
+    if(pft_distribution_method == pft_from_file .or. &
+         pft_distribution_method == pft_lowland_upland) then
+       do g = begg, endg
+          ! If hillslopes will be used in a gridcell, modify wt_nat_patch, otherwise use original patch distribution
+          if(ncolumns_hillslope(g) > 0) then
+             ! First patch gets 100% weight; all other natural patches are zeroed out
+             wt_nat_patch(g,:)         = 0._r8
+             wt_nat_patch(g,natpft_lb) = 100._r8
+          endif
+       enddo
+    endif
+
+    ! pft_uniform_dominant_pft uses the patch with the
+    ! largest weight for all hillslope columns in the gridcell
+    if (pft_distribution_method == pft_uniform_dominant_pft) then
+       do g = begg, endg
+          ! If hillslopes will be used in a gridcell, modify wt_nat_patch,
+          ! otherwise use original patch distribution
+          if(ncolumns_hillslope(g) > 0) then
+             call HillslopeDominantPftIndex(wt_nat_patch(g,:),natpft_lb,m)
+             wt_nat_patch(g,:) = 0._r8
+             wt_nat_patch(g,m) = 100._r8
+          endif
+       enddo
+    endif
+
+    ! pft_lowland_dominant_pft uses the two patches with the
+    ! largest weights for the hillslope columns in the gridcell
+    if (pft_distribution_method == pft_lowland_dominant_pft) then
+       do g = begg, endg
+          ! If hillslopes will be used in a gridcell, modify wt_nat_patch, otherwise use original patch distribution
+          if(ncolumns_hillslope(g) > 0) then
+             call HillslopeTwoLargestPftIndices(wt_nat_patch(g,:),natpft_lb,m,n)
+             ! Preserve the relative weights of the largest and
+             ! next largest weights using arbitrarily chosen values
+             ! (i.e. m should be larger than n)  This will minimize
+             ! memory usage while still allowing HillslopeDominantLowlandPft
+             ! to pick out the two largest patch types.
+             if(m /= n) then
```

Comment:
- [ ] What if `m == n`? It seems like there should be some handling of that case, even if for now it just means aborting with an error message. Or is this handled implicitly in some way that I'm not immediately seeing? (Note: If you change this to use `find_k_max_indices`, as I suggest in another comment, this logic would differ: I think the way that works is that it always returns `k` indices, but one of them may have a weight of 0 if there is only one PFT with a non-zero weight.)